File: Baekjoon/Python/BOJ_PS/Implementation/Q2174-robotSimulation.py
a, b = map(int, input().split())
n, m = map(int, input().split())

land = [[0] * a for _ in range(b)]

direction_dictionary = {
    'N': 0,
    'E': 1,
    'S': 2,
    'W': 3,
}

d = [
    [-1, 0],
    [0, 1],
    [1, 0],
    [0, -1]
]

robots = []
for i in range(n):
    x, y, direction = input().split()
    r, c = b - int(y), int(x) - 1
    robots.append([r, c, direction_dictionary[direction]])
    land[r][c] = i+1

def turn(current_d, command, count):
    add_number = 3 if command == "L" else 1
    for i in range(count):
        current_d = (current_d + add_number) % 4

    return current_d

is_wrong_command = False

for i in range(m):
    robot_number, command_kind, command_count = input().split()
    robot_number = int(robot_number)
    command_count = int(command_count)

    r, c, direction = robots[robot_number - 1]
    if command_kind == 'L' or command_kind == 'R':
        new_direction = turn(direction, command_kind, command_count)
        robots[robot_number - 1] = [r, c, new_direction]
    else:
        for j in range(0, command_count):
            nr = r + d[direction][0]
            nc = c + d[direction][1]

            if 0 <= nr < b and 0 <= nc < a:
                if land[nr][nc] > 0:
                    is_wrong_command = True
                    print("Robot", land[r][c], "crashes into robot", land[nr][nc])
                    break
                else:
                    land[r][c] = 0
                    land[nr][nc] = robot_number
                    robots[robot_number - 1] = [nr, nc, direction]
                    r, c = nr, nc
            else:
                is_wrong_command = True
                print("Robot", land[r][c], "crashes into the wall")
                break
    if is_wrong_command:
        break

if not is_wrong_command:
    print('OK')

File: Baekjoon/Swift/BOJ_PS/BOJ_PS/Implementation/Q2174-robotSimulation.swift
//
//  Q2174-robotSimulation.swift
//  BOJ_PS
//
//  Created by thoonk on 2022/07/21.
//

typealias Coord = (r: Int, c: Int)

let ab = readLine()!.split(separator: " ").map { Int(String($0))! }
let a = ab[0]
let b = ab[1]

let nm = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = nm[0]
let m = nm[1]

var land = [[Int]](repeating: [Int](repeating: 0, count: a), count: b)

// 0: x좌표, 1: y좌표, 2: 방향
var robots = [[Int]]()
//  0: 명령을 내리는 로봇, 1: 명령의 종류, 2: 명령의 반복 횟수
var commands = [[String](repeating: "", count: 3)]

var directionDictionary: [String: Int] = [
    "N": 0,
    "E": 1,
    "S": 2,
    "W": 3
]

let d = [(-1,0), (0,1), (1,0), (0,-1)]

for i in 0..<n {
    let robot = readLine()!.split(separator: " ").map { String($0) }
    
    let r = b - Int(robot[1])!
    let c = Int(robot[0])! - 1
    
    robots.append([r, c, directionDictionary[robot[2]]!])
    
    land[r][c] = i + 1
}


func turn(_ d: inout Int, _ command: String, _ count: Int) {
    let addNumber = command == "L" ? 3 : 1
    for _ in 0..<count {
        d = (d + addNumber) % 4
        print(d)
    }
}

var isWrongCommand = false

for _ in 0..<m {
    let command = readLine()!.split(separator: " ").map { String($0) }
    commands.append(command)
    
    let robotNumber = Int(command[0])!
    let commandKind = command[1]
    let commandCount = Int(command[2])!
    
    let robot = robots[robotNumber - 1]
    var r = robot[0]
    var c = robot[1]
    var direction = robot[2]

    if commandKind == "L" || commandKind == "R" {
        turn(&direction, commandKind, commandCount)
        robots[robotNumber - 1] = [r, c, direction]
    } else {
        for _ in 0..<commandCount {
            let nr = r + d[direction].0
            let nc = c + d[direction].1
            
            if 0 <= nr && nr < b && 0 <= nc && nc < a {
                if land[nr][nc] > 0 {
                    isWrongCommand = true
                    print("Robot \(land[r][c]) crashes into robot \(land[nr][nc])")
                    break
                } else {
                    land[r][c] = 0
                    land[nr][nc] = robotNumber
                    robots[robotNumber -  1] = [nr, nc, direction]
                    r = nr
                    c = nc
                }
            } else {
                isWrongCommand = true
                print("Robot \(land[r][c]) crashes into the wall")
                break
            }
        }
    }
    
    if isWrongCommand {
        break
    }
}

if !isWrongCommand {
    print("OK")
}

